### PR TITLE
SALTO-2803 - Salesforce exclude CustomObjects with many instances on fetch

### PR DESCRIPTION
**Pull request by idoser96**
When fetching salesforce, if a CustomObject has too many instances, exclude it and don't fetch them

---

This is the second PR of this issue, the first one was https://github.com/salto-io/salto/pull/3423

---
_Release Notes_: 
CustomObjects with many instances won't be fetched unless configured otherwise, which will avoid a slow fetching and a slow environment

---
_User Notifications_: 
On fetch, if there were CustomObjects that were added to the exclude list, the user will receive a message and will have to approve it in order to finish the fetch

### CODE REVIEW

File: packages/salesforce-adapter/test/adapter.discover.test.ts [hunk 1]
```diff
@@ -1176,25 +1176,63 @@ public class LargeClass${index} {
...
-      const { client: mockClient } = mockAdapter()
-      it('should not fetch the types and add them to the exclude list', async () => {
-        const filePropMock = mockFileProperties({ fullName: 'fullName', type: 'testType' })
-        const fetchResult = await fetchMetadataInstances({
-          client: mockClient,
-          metadataType: {} as unknown as ObjectType,
-          metadataQuery: {} as unknown as MetadataQuery,
-          fileProps: [filePropMock, filePropMock],
-          maxInstancesPerType: 1,
+      const metadataType = {
```
Comment (adamro) [inline packages/salesforce-adapter/test/adapter.discover.test.ts:1179]: This is not really related to this PR/ticket, right? It's a leftover from https://salto-io.atlassian.net/browse/SALTO-2236 ?
It should probably be in a separate PR/commit with the right ticket # connected to it

File: packages/salesforce-adapter/test/filters/custom_objects_instances.test.ts [hunk 2]
```diff
@@ -1048,6 +1068,54 @@ describe('Custom Object Instances filter', () => {
...
+      })
+    })
+    it('Should fetch CustomObjects with less instances than MaxInstancesPerType', async () => {
+      const elements = [testElement]
+      client.countInstances = jest.fn().mockResolvedValue(1)
+      const result = await filter.onFetch(elements)
+      expect(elements.length).toBe(2)
+      expect(result).toMatchObject({ configSuggestions: [] })
+    })
+  })
```
Comment (adamro) [inline packages/salesforce-adapter/test/filters/custom_objects_instances.test.ts:1118]: Can you also add the case where there are 2 types, one that has more than max and one under the max and see it doesn't fetch the one and does fetch the other please?

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts [hunk 3]
```diff
@@ -412,6 +417,33 @@ export const getCustomObjectsFetchSettings = async (
...
+    return undefined
+  }
+
+  const instancesCount = await client.countInstances(typeName)
+
+  if (instancesCount > maxInstancesPerType) {
+    // Remove the type from the fetch
+    delete fetchSettings[typeName]
+    // Return a config suggestion to exclude that type from the dataObjects
+    const reason = `'${typeName}' has ${instancesCount} instances so it was skipped and would be excluded from future fetch operations, as ${MAX_INSTANCES_PER_TYPE} is set to ${maxInstancesPerType}.
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/salesforce-adapter/src/filters/custom_objects_instances.ts:439]: Shouldn't because DataObjectExcludeChange is generic, but i've created a new function to handle this reason

[the remark this answers is not in the document]
Comment (idoser96) [reply packages/salesforce-adapter/src/filters/custom_objects_instances.ts:439]: createDataObjectExcludeChange is a generic function i created, so i don't think it should have any logic

also, i believe its better to keep the reason building inside the relevant function, so its clear what the function does

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts [hunk 4]
```diff
@@ -431,6 +463,19 @@ const filterCreator: RemoteFilterCreator = ({ client, config }) => ({
...
+    const filterHeavyTypesPromises = Object.keys(validChangesFetchSettings).map(
+      typeName => filterHeavyTypes({
+        typeName,
+        maxInstancesPerType: config.fetchProfile.maxInstancesPerType,
+        fetchSettings: validChangesFetchSettings,
+        client,
+      })
+    )
+    // Promise.all so they all run parallel
+    const heavyTypesConfigChanges = (await Promise.all(filterHeavyTypesPromises)).filter(isDefined)
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/salesforce-adapter/src/filters/custom_objects_instances.ts:477]: aren't client calls limited? (ClientRateLimitConfig)

File: packages/salesforce-adapter/test/adapter.discover.test.ts (hunk 1 above)
Comment (idoser96) [reply packages/salesforce-adapter/test/adapter.discover.test.ts:1201]: https://github.com/salto-io/salto/pull/3509

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts [hunk 5]
```diff
@@ -431,9 +463,25 @@ const filterCreator: RemoteFilterCreator = ({ client, config }) => ({
       validFetchSettings,
       setting => apiName(setting.objectType),
     )
+
+    const typeToInstancesCount: Record<string, number> = {}
+    const { maxInstancesPerType } = config.fetchProfile
+
+    await awu(Object.keys(validChangesFetchSettings)).forEach(async typeName => {
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/custom_objects_instances.ts:470]: Why `forEach` and not `map`?
return [typeName, count] and then do a sync Object.fromEntries

Also, wouldn't it make more sense to not run the code at all if it's `UNLIMITED_INSTANCES_VALUE`?

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamro) [reply packages/salesforce-adapter/src/filters/custom_objects_instances.ts]: But if you look at the other functions in `createDataObjectExcludeChange`'s file, the logic of creating the reason based on the relevant inputs and the place that has the relevant text is there, not where it's called.
So to keep that convention I would expect `createDataObjectExcludeChange` to get `typeName, instanceCount and maxInstancesPerType` at the function params and have the reason creation be there.
Currently if someone wants to look at the configSuggestion's texts and review them they are located as the place they are created, you are now adding an extra place which is the creator, only for this specific case.

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts [hunk 6]
```diff
@@ -431,9 +463,25 @@ const filterCreator: RemoteFilterCreator = ({ client, config }) => ({
...
+      typeToInstancesCount[typeName] = maxInstancesPerType === UNLIMITED_INSTANCES_VALUE
+        ? UNLIMITED_INSTANCES_VALUE
+        : await client.countInstances(typeName)
+    })
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/custom_objects_instances.ts:474]: Once you have a map of only typeName->instancesCount for the ones above max (comment above).
You have 2 things to do - 
1. Filter out the relevant typeNames from `validChangesFetchSettings` (Basically use _.omit with the keys of the above map - the "maxed" typeNames)
2. Create the configSuggestions. Run a .map on the entries of our typeName->instancesCount that runs  `createDataObjectExcludeChange(typeName, typeName, instancesCount, maxInstancesPerType)` if you go with my suggestion for `createDataObjectExcludeChange`

That would be much more in-style with our code-base with the current `for` implementation you have in `filterTypesWithManyInstances`

File: packages/salesforce-adapter/src/client/client.ts [hunk 7]
```diff
@@ -503,6 +503,17 @@ export default class SalesforceClient {
     return requestWithRetry(this.retryOptions.maxAttempts ?? DEFAULT_RETRY_OPTS.maxAttempts)
   }
 
+  /**
+   * Count the amount of instances of a type
+   */
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/client/client.ts:508]: Nit: I think the code is self explanitory.. No need for this documentation.

**Review by amitgoldarb — COMMENTED** [review]
Looks great!
Added some small comments.

File: packages/salesforce-adapter/src/config_change.ts [hunk 8]
```diff
@@ -41,6 +51,21 @@ export const getConfigChangeMessage = (configChanges: ConfigChangeSuggestion[]):
   return [MESSAGE_INTRO, '', MESSAGE_REASONS_INTRO, ...reasons.map(formatReason), '', MESSAGE_SUMMARY].join('\n')
 }
 
+export const createManyInstancesExcludeChange = (
```
Comment (adamro) [inline packages/salesforce-adapter/src/config_change.ts:54]: naming wise to be aligned with the rest of the file I think this should be
`createManyInstancesExcludeConfigChange` :\ (the sad face is cause this is long, but still)

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts [hunk 9]
```diff
@@ -445,7 +493,11 @@ const filterCreator: RemoteFilterCreator = ({ client, config }) => ({
         ))
       .toArray()
     return {
-      configSuggestions: [...invalidFieldSuggestions, ...configChangeSuggestions],
+      configSuggestions: [
+        ...invalidFieldSuggestions,
+        ...heavyTypesConfigChanges,
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/custom_objects_instances.ts:498]: naming wise I guess this should be `heavyTypesSuggestions` to be aligned with the others

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts [hunk 10]
```diff
@@ -412,6 +415,41 @@ export const getCustomObjectsFetchSettings = async (
     .toArray()
 }
 
+const filterTypesWithManyInstances = async (
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/custom_objects_instances.ts:418]: I still feel this is not inline with our general coding style guideline and would expect this to be maps & filters, not a forEach that builds external data structures, but considering it's style and I am repeating myself hence didn't convince you I am not gonna insist

**Review by adamro — APPROVED** [review]
Notice there are still a couple of naming comments but feel free to merge after to save a cycle

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts (hunk 10 above)
Comment (idoser96) [reply packages/salesforce-adapter/src/filters/custom_objects_instances.ts:418]: 👍 thanks
